document-store/src/main/java/org/hypertrace/core/documentstore/query/GenericQuery.java

```diff
@@ -0,0 +1,95 @@
+package org.hypertrace.core.documentstore.query;
+
+import java.util.ArrayList;
+import java.util.List;
+import lombok.Builder;
+import lombok.Singular;
+import lombok.Value;
+import org.hypertrace.core.documentstore.expression.Filterable;
+import org.hypertrace.core.documentstore.expression.Groupable;
+import org.hypertrace.core.documentstore.expression.Projectable;
+
+/**
+ * A generic query definition that supports expressions. Note that this class is a more general
+ * version of {@link org.hypertrace.core.documentstore.Query}
+ *
+ * <p>Example: <code>
+ *     SELECT col4, SUM(col5) AS total
+ *     FROM <collection>
+ *     WHERE col1 < 7 AND col2 != col3
+ *     GROUP BY col4, col6
+ *     HAVING SUM(col5) >= 100
+ *     ORDER BY col7+col8 DESC
+ *     OFFSET 5
+ *     LIMIT 10
+ * </code> can be built as <code>
+ *     GenericQuery query = GenericQuery.builder()
+ *         .projection(LiteralExpression.of("col4"))
+ *         .projection(
+ *             AggregateExpression.of(SUM, LiteralExpression.of("col5")),
+ *             "total")
+ *         .whereFilter(LogicalExpression.of(
+ *             RelationalExpression.of(
+ *                 LiteralExpression.of("col1"),
+ *                 LT,
+ *                 ConstantExpression.of(7)),
+ *             AND,
+ *             RelationalExpression.of(
+ *             LiteralExpression.of("col2"),
+ *             NEQ,
+ *             LiteralExpression.of("col3"))))
+ *         .groupBy(LiteralExpression.of("col4"))
+ *         .groupBy(LiteralExpression.of("col6"))
+ *         .havingFilter(
+ *             RelationalExpression.of(
+ *                 AggregateExpression.of(SUM, LiteralExpression.of("col5")),
+ *                 GTE,
+ *                 ConstantExpression.of(100)))
+ *         .orderBy(
+ *             GenericOrderBy.of(
+ *                 ArithmeticExpression.builder()
+ *                     .operand(LiteralExpression.of("col7"))
+ *                     .operator(ADD)
+ *                     .operand(LiteralExpression.of("col8")).build(),
+ *                 false))
+ *         .offset(5)
+ *         .limit(10)
+ *         .build();
+ *  </code>
+ */
+@Value
+@Builder
+public class GenericQuery {
+  List<Projection> projections;
+  Filterable whereFilter;
+
+  @Singular List<Groupable> groupBys;
+  Filterable havingFilter;
+
+  @Singular List<GenericOrderBy> orderBys;
+
+  int offset;
```

Comment:
I don't know what our existing support looks like - do we normally work in cursors, offsets or both?

Reply:
I see that the existing implementation exposes offsets outside (and works with cursors internally).